language/vm/src/compatibility.rs

```diff
@@ -86,13 +96,81 @@ impl Compatibility {
+        let mut old_api_usage_by_friend_modules = HashMap::new();
+        for friend_module in old_friends {
+            for func_handle in friend_module.function_handles() {
+                if friend_module
+                    .module_id_for_handle(friend_module.module_handle_at(func_handle.module))
+                    != module_id
+                {
+                    continue;
+                }
+                let func_sig = FunctionSignature::new(friend_module, func_handle);
+                old_api_usage_by_friend_modules
+                    .entry(friend_module.self_id())
+                    .or_insert_with(HashSet::new)
+                    .insert(func_sig);
+            }
+        }
+
+        // The modules are considered as compatible function-wise when all the conditions are met:
         //
-        // TODO: the current implementation treats `public(friend)` function the same way as`public`
-        // for compatibility checking. We might need to change this in the future to allow
-        // `public(friend)` functions to update if none of its friends actually calls it.
-        for item in &old_module.externally_visible_functions {
-            if !new_module.externally_visible_functions.contains(&item) {
+        // - old module's public functions are a subset of the new module's public functions
+        //   (i.e. we cannot remove or change public functions)
+        // - old module's script functions are a subset of the new module's script functions
+        //   (i.e. we cannot remove or change script functions)
+        // - for any friend function that is removed or changed in the old module
+        //   - if the function visibility is upgraded to public, it is OK
+        //   - otherwise, it must not be used by any module in the old friend-list.
+        //     Due to the current publish-one-module-at-a-time model, we know that there is no
+        //     module calling the new friend function.
+        for (func_sig, old_vis) in &old_module.exposed_functions {
+            let new_vis_opt = new_module.exposed_functions.get(func_sig);
+            let is_compatible = match (old_vis, new_vis_opt) {
+                (Visibility::Public, Some(Visibility::Public)) => true,
+                (Visibility::Public, _) => false,
+                (Visibility::Script, Some(Visibility::Script)) => true,
+                (Visibility::Script, _) => false,
+                (Visibility::Friend, Some(Visibility::Public)) => true,
+                (Visibility::Friend, Some(Visibility::Friend)) => true,
+                (Visibility::Friend, Some(Visibility::Script))
+                | (Visibility::Friend, Some(Visibility::Private))
+                | (Visibility::Friend, None) => !old_api_usage_by_friend_modules
+                    .values()
+                    .any(|usage| usage.contains(func_sig)),
+                (Visibility::Private, _) => unreachable!("A private function can never be exposed"),
```

Comment:
For paranoia, we might want `exposed_functions` to be a `BTreeMap`. For example: if we have a bug that triggers this case, one validator might hit a `false` first, but another but hit this first `unreachable` first. That would lead them to disagree on whether the module upgrade should be aborted (if `false` is hit) or discarded. 

By contrast, if we used a `BTreeMap`, they would agree on the outcome. This is an extreme case because it relies on introducing another bug to even see the problem, but it doesn't hurt to be defensive, particularly if we haven't observed that `BTreeMap` causes perf problems that are fixed by using a `HashMap`.

Reply:
Totally agreed. I'll make an update. My bad for not using the `BTreeMap` in the first place, don't know what I was thinking...

Reply:
Updated to `BTreeMap` (and also one `HashSet` -> `BTreeSet`).